src/ui/VirtualFileRenderer/VirtualFileRenderer.spec.tsx

```diff
@@ -0,0 +1,361 @@
+import { fireEvent, render, screen, waitFor } from '@testing-library/react'
+import { userEvent } from '@testing-library/user-event'
+// eslint-disable-next-line no-restricted-imports
+import { type Dictionary } from 'lodash'
+import { MemoryRouter, Route, useLocation } from 'react-router-dom'
+
+import { VirtualFileRenderer } from './VirtualFileRenderer'
+
+window.requestAnimationFrame = (cb) => {
+  cb(1)
+  return 1
+}
+window.cancelAnimationFrame = () => {}
+
+const scrollToMock = jest.fn()
+window.scrollTo = scrollToMock
+
+class ResizeObserver {
+  callback = (x: any) => null
+
+  constructor(callback: any) {
+    this.callback = callback
+  }
+
+  observe() {
+    this.callback([{ target: { getAttribute: () => ({ scrollWidth: 100 }) } }])
+  }
+  unobserve() {
+    // do nothing
+  }
+  disconnect() {
+    // do nothing
+  }
+}
+global.window.ResizeObserver = ResizeObserver
+
+const code = `<Breadcrumb
+    paths={[
+    { pageName: 'owner', text: owner },
+    { pageName: 'repo', text: repo },
+    ...treePaths,
+    {..props}
+    ]}
+/>`
+
+const coverageData = {
+  0: 'H',
+  1: 'M',
+  2: 'P',
+  3: 'H',
+  4: 'M',
+  5: 'P',
+} as unknown as Dictionary<'H' | 'M' | 'P'>
+
+let testLocation: ReturnType<typeof useLocation>
+const wrapper =
+  (initialEntry = '/'): React.FC<React.PropsWithChildren> =>
+  ({ children }) => (
+    <MemoryRouter initialEntries={[initialEntry]}>
+      <Route path="/">{children}</Route>
+      <Route
+        path="*"
+        render={({ location }) => {
+          testLocation = location
+          return null
+        }}
+      />
+    </MemoryRouter>
+  )
+
+describe('VirtualFileRenderer', () => {
+  function setup() {
+    const user = userEvent.setup()
+
+    return { user }
+  }
+
+  it('renders the code', () => {
+    render(
+      <VirtualFileRenderer
+        code={code}
+        coverage={coverageData}
+        fileName="tsx"
+      />,
+      {
+        wrapper: wrapper(),
+      }
+    )
+
+    const codeBlock = screen.getByText(/<Breadcrumb/)
+    expect(codeBlock).toBeInTheDocument()
+  })
+
+  it('renders line numbers', () => {
+    render(
+      <VirtualFileRenderer
+        code={code}
+        coverage={coverageData}
+        fileName="tsx"
+      />,
+      {
+        wrapper: wrapper(),
+      }
+    )
+
+    const lineNumbers = screen.getAllByText(/\d+/)
+    expect(lineNumbers).toHaveLength(8)
+  })
+
+  describe('covered lines', () => {
+    it('uses the correct color', () => {
+      render(
+        <VirtualFileRenderer
+          code={code}
+          coverage={coverageData}
+          fileName="tsx"
+        />,
+        {
+          wrapper: wrapper(),
+        }
+      )
+
+      const coveredLine = screen.getByText(0)
+      expect(coveredLine).toBeInTheDocument()
```

Comment:
Unless this line performs an assertion, then I think you can remove it as the line after will fail if its not in the document